src/proxy/h2.rs

```diff
@@ -71,21 +71,41 @@ async fn do_ping_pong(
 
 // H2Stream represents an active HTTP2 stream. Consumers can only Read/Write
 pub struct H2Stream {
-    pub read: H2StreamReadHalf,
-    pub write: H2StreamWriteHalf,
+    read: H2StreamReadHalf,
+    write: H2StreamWriteHalf,
 }
 
 pub struct H2StreamReadHalf {
-    pub recv_stream: h2::RecvStream,
-    pub buf: Bytes,
-    pub half_dropped: Arc<()>,
-    pub active_count: Arc<AtomicU16>,
+    recv_stream: h2::RecvStream,
+    buf: Bytes,
+    _dropped: Option<DropCounter>,
 }
 
 pub struct H2StreamWriteHalf {
-    pub send_stream: h2::SendStream<SendBuf>,
-    pub half_dropped: Arc<()>,
-    pub active_count: Arc<AtomicU16>,
+    send_stream: h2::SendStream<SendBuf>,
+    _dropped: Option<DropCounter>,
+}
+
+struct DropCounter {
+    // Whether the other end of this shared counter has already dropped.
+    // We only decrement if they have, so we do not double count
+    half_dropped: Arc<()>,
```

Comment:
Could we just use an AtomicBool here?

Reply:
Yeah this is basically the same thing but it makes me feel smart :slightly_smiling_face:

Reply:
I think the reason I didn't do this was it would need to be Arc<AtomicBool>. Would work though

Reply:
If it's gotta be Arc'd anyway then nevermind yea